### PR TITLE
Move call epoch_step() method to begin of epoch.

### PR DESCRIPTION
**Pull request by asenina**
- Moved calls `epoch_step()` and `step()` to begin of epoch.

-  Moved `sparsity_init` and `pruning_init` params to the logic of the algorithm. Now the initial sparsity and pruning levels are set in `create_compressed_model` .

### CODE REVIEW

File: examples/classification/configs/sparsity_quantization/mobilenet_v2_imagenet_rb_sparsity_int8.json [hunk 1]
```diff
@@ -27,8 +27,8 @@
     "compression": [
         {
             "algorithm": "rb_sparsity", // Compression algorithm name
+            "sparsity_init": 0.01, // Inital sparsity ratio, e.g. value "0.1" means that the method sets 10% of zero weights as a target after the training process starts
             "params": {
-                "sparsity_init": 0.01, // Inital sparsity ratio, e.g. value "0.1" means that the method sets 10% of zero weights as a target after the training process starts
                 "sparsity_target": 0.52, // Desired sparsity ratio, e.g. value "0.5" means that the method will schedule the training so that to get 50% of zero weights in the end
                 "sparsity_target_epoch": 5, // Number of epochs at which sparsity ratio will be increased from "sparsity_init" value up to "sparsity_target" value
                 "sparsity_freeze_epoch": 10 // Overall number of epochs that are used to train sparsity mask
```
Comment (gadylshintr) [inline examples/classification/configs/sparsity_quantization/mobilenet_v2_imagenet_rb_sparsity_int8.json:34]: As @vshampor stated we've  decided to interpret `sparsity_target_epoch` and `sparsity_freeze_epoch` as epoch index.

File: nncf/sparsity/schedulers.py [hunk 2]
```diff
@@ -84,8 +84,7 @@ def __init__(self, sparsity_algo, params=None):
                                "but steps_per_epoch was not set in config. Will only start updating "
                                "sparsity level after measuring the actual steps per epoch as signaled "
                                "by a .epoch_step() call.")
-
-        self._set_sparsity_level()
+        #self._set_sparsity_level()
```
Comment (gadylshintr) [inline nncf/sparsity/schedulers.py:87]: Is that will be removed or it means that it could be optionally called (why if so)?

File: nncf/compression_method_api.py [hunk 3]
```diff
@@ -100,6 +104,9 @@ def state_dict(self):
     def initialize(self):
         pass
 
+    @property
+    def current_epoch(self):
```
Comment (gadylshintr) [inline nncf/compression_method_api.py:108]: As we use `sparsity_init` to initialize algorithm directly and so we don't call set_sparsity in `CompressionScheduler` constructor it seems that we do not need `current_epoch` and `_current_epoch` anymore - `last_epoch` should be enough.

File: examples/object_detection/main.py [hunk 4]
```diff
@@ -290,7 +290,10 @@ def train(net, compression_ctrl, train_data_loader, test_data_loader, criterion,
     best_compression_level = CompressionLevel.NONE
     test_freq_in_epochs = max(config.test_interval // epoch_size, 1)
 
+    last_epoch = -1 if config.start_iter == 0 else config.start_iter // epoch_size
+    compression_ctrl.scheduler.epoch_step(last_epoch)
```
Comment (gadylshintr) [inline examples/object_detection/main.py:294]: Why do we set `last_epoch` to -1 if we call then `compression_ctrl.scheduler.epoch_step(last_epoch)` which doesn't  increment it.

File: examples/object_detection/main.py [hunk 5]
```diff
@@ -290,7 +290,10 @@ def train(net, compression_ctrl, train_data_loader, test_data_loader, criterion,
...
     for iteration in range(config.start_iter, config['max_iter']):
+        compression_ctrl.scheduler.step(iteration - config.start_iter)
```
Comment (gadylshintr) [inline examples/object_detection/main.py:296]: Shouldn't it be this way (without other calls of epoch_step):
```
epoch = iteration // epoch_size
if iteration % epoch_size == 0:
    compression_ctrl.scheduler.epoch_step(epoch) # or without epoch, as for resuming we've already loaded scheduler state from checkpoint
compression_ctrl.scheduler.step(iteration) # or without iteration
```
It also looks like currently it won't work correctly for resuming from intermediate step (i.e. in a middle of epoch) with per-step scheduler. This refers to scheduler implementation rather than to this sample.

File: nncf/sparsity/schedulers.py (hunk 2 above)
Comment (asenina) [reply nncf/sparsity/schedulers.py:87]: It will be removed.

File: examples/object_detection/main.py (hunk 5 above)
Comment (asenina) [reply examples/object_detection/main.py]: Done.

File: examples/object_detection/main.py (hunk 4 above)
Comment (asenina) [reply examples/object_detection/main.py]: Fixed.

**Comment by asenina** [comment]
please take a look  @gadylshintr @vshampor

File: nncf/compression_method_api.py (hunk 3 above)
Comment (vshampor) [reply nncf/compression_method_api.py:110]: @gadylshintr I think that thinking in `current_epoch` terms is more straightforward.

File: examples/classification/configs/sparsity_quantization/mobilenet_v2_imagenet_rb_sparsity_int8.json (hunk 1 above)
Comment (vshampor) [reply examples/classification/configs/sparsity_quantization/mobilenet_v2_imagenet_rb_sparsity_int8.json:34]: Should probably remove these in-place comments altogether - we have readmes with the same kind of documentation already.

File: nncf/compression_method_api.py (hunk 3 above)
Comment (gadylshintr) [reply nncf/compression_method_api.py:110]: Okay. It is probably should be the same change to `last_step` for consistency?

File: nncf/compression_method_api.py [hunk 6]
```diff
@@ -86,31 +85,30 @@ def step(self, last=None):
...
         """
-        Should be called after each training epoch.
+        Should be called before each training epoch.
         Arguments:
             `last` - specifies the initial "previous" epoch
         """
         if last is None:
-            last = self.last_epoch + 1
             self._current_epoch += 1
+            last = self._current_epoch - 1
```
Comment (gadylshintr) [inline nncf/compression_method_api.py:94]: Unused variable?

File: nncf/compression_method_api.py (hunk 3 above)
Comment (vshampor) [reply nncf/compression_method_api.py:110]: I agree, @asenina please modify the code accordingly

File: examples/object_detection/main.py [hunk 7]
```diff
@@ -298,17 +298,15 @@ def train(net, compression_ctrl, train_data_loader, test_data_loader, criterion,
...
     for iteration in range(config.start_iter, config['max_iter']):
-        compression_ctrl.scheduler.step(iteration - config.start_iter)
         if (not batch_iterator) or (iteration % epoch_size == 0):
             # create batch iterator
             batch_iterator = iter(train_data_loader)
 
         epoch = iteration // epoch_size
 
-        if (iteration + 1) % epoch_size == 0:
+        compression_ctrl.scheduler.step(iteration - config.start_iter)
```
Comment (gadylshintr) [inline examples/object_detection/main.py:308]: Why argument is not just `iteration`?

File: nncf/config_schema.py [hunk 8]
```diff
@@ -475,9 +475,6 @@ def with_attributes(schema: Dict, **kwargs) -> Dict:
     "concave": with_attributes(_BOOLEAN, description="For polynomial scheduler - if True, then the target sparsity "
                                                      "level will be approached in concave manner, and in convex "
                                                      "manner otherwise."),
-    #"sparsity_init": with_attributes(_NUMBER,
-    #                                 description="Initial value of the sparsity level applied to the "
-    #                                             "model"),
     "sparsity_target": with_attributes(_NUMBER,
                                        description="Target value of the sparsity level for the model"),
     "sparsity_target_epoch": with_attributes(_NUMBER,
```
Comment (gadylshintr) [inline nncf/config_schema.py:480]: Reformulate, please, in terms of epoch indexes.

Comment (gadylshintr) [inline nncf/config_schema.py:480]: Reformulate description in terms of epoch indexes, please. (and for `sparsity_freeze_epoch` too)

File: nncf/sparsity/schedulers.py [hunk 9]
```diff
@@ -117,14 +115,13 @@ def load_state_dict(self, state_dict):
 
     @property
     def current_sparsity_level(self):
```
Comment (gadylshintr) [inline nncf/sparsity/schedulers.py:117]: If `sparsity_target_epoch` is 0, sparsity should be equal to `target_sparsity` (for exponential too)

File: examples/classification/configs/sparsity_quantization/mobilenet_v2_imagenet_rb_sparsity_int8.json (hunk 1 above)
Comment (asenina) [reply examples/classification/configs/sparsity_quantization/mobilenet_v2_imagenet_rb_sparsity_int8.json]: Done.

File: nncf/compression_method_api.py (hunk 3 above)
Comment (asenina) [reply nncf/compression_method_api.py:107]: Done.

File: nncf/compression_method_api.py (hunk 6 above)
Comment (asenina) [reply nncf/compression_method_api.py]: Removed.

File: nncf/config_schema.py (hunk 8 above)
Comment (asenina) [reply nncf/config_schema.py:480]: Done.

File: nncf/sparsity/schedulers.py (hunk 9 above)
Comment (asenina) [reply nncf/sparsity/schedulers.py:117]: Done.

File: docs/compression_algorithms/Sparsity.md [hunk 10]
```diff
@@ -81,13 +81,13 @@ The magnitude sparsity method implements a naive approach that is based on the a
...
     }
+    "sparsity_init": 0.05,// "Initial value of the sparsity level applied to the model in 'create_compressed_model' function
     "params": {
             "schedule": "multistep",  // The type of scheduling to use for adjusting the target sparsity level
             "patience": 3, // A regular patience parameter for the scheduler, as for any other standard scheduler. Specified in units of scheduler steps.
-            "sparsity_init": 0.05,// "Initial value of the sparsity level applied to the model
             "sparsity_target": 0.7, // Target value of the sparsity level for the model
-            "sparsity_target_epoch": 3, // The target sparsity value will be reached after this many epoch steps
-            "sparsity_freeze_epoch": 50, // The number of steps after which the sparsity mask will be frozen and no longer trained
+            "sparsity_target_epoch": 3, // Index of the epoch from which the sparsity level of the model will be equal to spatsity_target value
```
Comment (gadylshintr) [inline docs/compression_algorithms/Sparsity.md:89]: 1. I would like to clarify how should it work in case of per-step sparsity scheduler: will sparsity rich target value at the first step of `sparsity_target_epoch` or at the last step, @vshampor what do you think?

File: docs/compression_algorithms/Sparsity.md [hunk 11]
```diff
@@ -81,13 +81,13 @@ The magnitude sparsity method implements a naive approach that is based on the a
...
+            "sparsity_freeze_epoch": 50, // // Index of the epoch from which the sparsity mask will be frozen and no longer trained
```
Comment (gadylshintr) [inline docs/compression_algorithms/Sparsity.md:90]: Extra //

File: examples/object_detection/main.py [hunk 12]
```diff
@@ -305,7 +305,8 @@ def train(net, compression_ctrl, train_data_loader, test_data_loader, criterion,
...
+        if iteration % epoch_size == 0:
```
Comment (gadylshintr) [inline examples/object_detection/main.py:309]: Not all block should be under `if iteration % epoch_size == 0`, some should be under `if iteration + 1 % epoch_size == 0`

File: docs/compression_algorithms/Sparsity.md (hunk 10 above)
Comment (vshampor) [reply docs/compression_algorithms/Sparsity.md:89]: Should probably later make `*_target_step` and `*_target_epoch` settings mutually exclusive. The step is calculated globally, across all epochs. I'm not sure that we even have a `*_target_step` option.

Comment (gadylshintr) [reply docs/compression_algorithms/Sparsity.md:89]: I'm not sure whether we need `*_target_step` or not it looks like new feature, but for `*_target_epoch` (for per-step case) I would like to clarify which training step of `*_target_epoch` epoch will be trained with target sparsity value the first or the last. It seems to me that it should be the last step of `*_target_epoch`. If we agree on that lets state it?

**Comment by asenina** [comment]
Jenkins please retry a build

File: examples/object_detection/main.py (hunk 12 above)
Comment (asenina) [reply examples/object_detection/main.py:309]: Fixed.

File: docs/compression_algorithms/Sparsity.md (hunk 11 above)
Comment (asenina) [reply docs/compression_algorithms/Sparsity.md]: Fixed. Thanks)

File: examples/object_detection/main.py (hunk 7 above)
Comment (asenina) [reply examples/object_detection/main.py]: Fixed.

**Review by vshampor — APPROVED** [review]

**Review by gadylshintr — COMMENTED** [review]
Have no more opened discussions
